Add distributed data parallel benchmark tool

pietern opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30660 Add distributed data parallel benchmark tool**

The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

Differential Revision: [D18781560](https://our.internmc.facebook.com/intern/diff/D18781560)

pietern committed:
Add distributed data parallel benchmark tool

The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

[ghstack-poisoned]

pietern committed:
Update on "Add distributed data parallel benchmark tool"


The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

[ghstack-poisoned]

pietern committed:
Update on "Add distributed data parallel benchmark tool"


The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

[ghstack-poisoned]

pietern commented:
Example output of the diff tool:

```
$ python3 diff.py PATH_TO_BASELINE_FILE PATH_TO_TEST_FILE
                                 baseline                      test
                     --------------------      --------------------
cuda_version:                        10.0  vs                  10.0
pytorch_version:          1.4.0a0+fcb7371  vs       1.4.0a0+fcb7371
Benchmark: resnet50 with batch size 32
   1 GPUs: p75:  0.097 ( -0.0%)  p95:  0.097 ( +0.1%)
   2 GPUs: p75:  0.104 ( -0.8%)  p95:  0.104 ( -0.8%)
   4 GPUs: p75:  0.105 ( -1.3%)  p95:  0.105 ( -1.4%)
   8 GPUs: p75:  0.108 ( -3.5%)  p95:  0.108 ( -3.5%)
  16 GPUs: p75:  0.111 ( -5.4%)  p95:  0.113 ( -7.6%)
  24 GPUs: p75:  0.111 ( -5.4%)  p95:  0.112 ( -5.9%)
  32 GPUs: p75:  0.112 ( -5.5%)  p95:  0.114 ( -7.8%)
  40 GPUs: p75:  0.114 ( -7.3%)  p95:  0.115 ( -7.9%)
  48 GPUs: p75:  0.114 ( -7.1%)  p95:  0.118 (-10.0%)
  56 GPUs: p75:  0.117 ( -9.4%)  p95:  0.118 ( -9.9%)
  64 GPUs: p75:  0.113 ( -6.4%)  p95:  0.114 ( -5.2%)
```

zhaojuanmao commented on benchmarks/ddp/README.md:
```diff
@@ -0,0 +1,112 @@
+
+
+Benchmark: resnet50 with batch size 32
+   1 GPUs --   warmup: p50:  0.098   327/s  p75:  0.098   326/s  p90:  0.098   326/s  p95:  0.098   326/s
```
it would be appreciated to explain the metric is for per GPU metric, and what does "0.098" mean and what does "327/s" mean?

pritamdamania87 replied:
Yes, column headers would be nice for this output. Also, would be nice to mention the unit for latency (ex: 0.098)

pietern replied:
Updated:

```
                            sec/iter    ex/sec      sec/iter    ex/sec      sec/iter    ex/sec      sec/iter    ex/sec
   1 GPUs --   warmup:  p50:  0.097s     328/s  p75:  0.097s     328/s  p90:  0.098s     327/s  p95:  0.098s     327/s
   1 GPUs --    1M/1G:  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     328/s
   2 GPUs --    1M/2G:  p50:  0.103s     310/s  p75:  0.103s     309/s  p90:  0.103s     309/s  p95:  0.104s     307/s
   4 GPUs --    1M/4G:  p50:  0.103s     310/s  p75:  0.103s     310/s  p90:  0.103s     310/s  p95:  0.103s     310/s
   8 GPUs --    1M/8G:  p50:  0.104s     307/s  p75:  0.104s     306/s  p90:  0.104s     306/s  p95:  0.104s     306/s
  16 GPUs --    2M/8G:  p50:  0.104s     306/s  p75:  0.105s     306/s  p90:  0.105s     305/s  p95:  0.106s     302/s
```

zhaojuanmao commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+#
+# This program performs a sweep over a) a number of model architectures, and
+# b) an increasing number of processes. This produces a 1-GPU baseline,
+# an 8-GPU baseline (if applicable), as well as measurements for however
```
nit: how many

pietern replied:
Intentional (see https://www.merriam-webster.com/dictionary/however for "however many").

pritamdamania87 commented on benchmarks/ddp/README.md:
```diff
@@ -0,0 +1,112 @@
+pytorch_version:          1.4.0a0+fcb7371  vs       1.4.0a0+fcb7371
+
+Benchmark: resnet50 with batch size 32
+   1 GPUs: p75:  0.097 ( -0.0%)  p95:  0.097 ( +0.1%)
```
Is it possible to diff the throughput as well? Also, column headers would be nice in this output as well.

pietern replied:
Yep, added throughput and headers.

pietern replied:
Yep, added throughput and header to the diff tool as well.

pritamdamania87 commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+    args = parser.parse_args()
+
+    num_gpus_per_node = torch.cuda.device_count()
+    assert num_gpus_per_node == 8, "Expected 8 GPUs per machine"
```
Is it possible to remove this limitation and use more or less GPUs based on the machine configuration?

pietern replied:
It is, but will require a bit more work to figure out which GPU to use on any given rank.

I've only run this yet on 8 GPU machines and think we can save this for a later PR.

pritamdamania87 commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+    assert num_gpus_per_node == 8, "Expected 8 GPUs per machine"
+
+    dist.init_process_group(
+        backend="gloo",
```
Maybe we can add a comment here mentioning this process_group is just used for initialization and we don't actually use this for the actual benchmark.

pietern replied:
Done.

pritamdamania87 commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+        benchmarks.append(TorchvisionBenchmark(device=device, model=args.model))
+    else:
+        for model in ["resnet50", "resnet101", "resnext50_32x4d", "resnext101_32x8d"]:
+            benchmarks.append(TorchvisionBenchmark(device=device, model=model))
```
Should we provide the ability for users to specify a model file path that we can load to run the benchmark?

pietern replied:
We could do so in a follow up. I don't know what it would entail here, so I'd rather merge this one first before adding more features.

pritamdamania87 commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+        local_print("\n")
+
+    group_nccl = dist.new_group(ranks=ranks, backend="nccl")
+    group_gloo = dist.new_group(ranks=ranks, backend="gloo")
```
Why do we have group_gloo here? Also, it seems like the backend should be configurable via a command line arg to this script.

pietern replied:
Good point. I've made it configurable.

pritamdamania87 commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+    # if the PyTorch 1.0 DistributedDataParallel class with Python
+    # side reduction is used. If we don't explicitly run GC,
+    # creation of NCCL process groups later on results in a crash.
+    gc.collect()
```
This seems like a bug? Is there a gh issue explaining what fails if you initialize NCCL process groups without calling gc.collect?

pietern replied:
Copied this code from an earlier version of this benchmark script I had. I tested with the hack removed and all works fine, so it isn't needed anymore. Removed the whole chunk.

pritamdamania87 commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+
+def benchmark_process_group(pg, benchmark):
+    torch.manual_seed(pg.rank())
+    torch.cuda.manual_seed(pg.rank())
```
Any reason for not having the same seed across all ranks?

pietern replied:
Different inputs to the model, in case there is data dependent control flow.

pritamdamania87 commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+                          0.001,
+                          momentum=0.9,
+                          weight_decay=1e-4)
+    if pg.size() > 1:
```
Is it possible to use DDP even for pg.size() == 1? This way all the results exercise the same code path.

pietern replied:
It is possible (I tried it). But the result is surprisingly off by a couple percent, indicating there is non-trivial overhead in DDP itself. I'm adding another option that adds a test run for 1 GPU with/without DDP, so that we always measure the impact. Already a good find!

pritamdamania87 commented on benchmarks/ddp/benchmark.py:
```diff
@@ -0,0 +1,276 @@
+        benchmarks.append((prefix, ranks))
+
+    # Every process runs once by themselves to warm up (CUDA init, etc).
+    append_benchmark("  warmup", [dist.get_rank()])
```
Do we need this warmup benchmark? Don't we already have warmup iterations in `benchmark_process_group`?

pietern replied:
This full warmup iteration is to make sure every rank has performed the same warmup, before starting the real ranks. Otherwise, we'd progressively need to warm up more ranks.

pietern commented:
New output of the diff tool:

```
$ ./diff.py result/nccl.json result/gloo.json
                                 baseline                      test
                     --------------------      --------------------
cuda_version:                        10.0  vs                  10.0
distributed_backend:                 nccl  vs                  gloo
pytorch_version:          1.4.0a0+05140f0  vs       1.4.0a0+05140f0

Benchmark: resnet50 with batch size 32

                  sec/iter    ex/sec      diff        sec/iter    ex/sec      diff
   1 GPUs:  p75:    0.109s     293/s     -8.6%  p95:    0.110s     291/s     -9.4%
   2 GPUs:  p75:    0.121s     264/s    -17.5%  p95:    0.121s     263/s    -17.6%
   4 GPUs:  p75:    0.138s     232/s    -33.8%  p95:    0.144s     222/s    -39.2%
   8 GPUs:  p75:    0.211s     151/s   -102.4%  p95:    0.222s     143/s   -113.3%
  16 GPUs:  p75:    0.256s     124/s   -145.8%  p95:    0.266s     120/s   -155.1%

Benchmark: resnet101 with batch size 32

                  sec/iter    ex/sec      diff        sec/iter    ex/sec      diff
   1 GPUs:  p75:    0.182s     175/s     -6.5%  p95:    0.182s     175/s     -5.8%
   2 GPUs:  p75:    0.198s     161/s    -12.6%  p95:    0.199s     160/s    -13.0%
   4 GPUs:  p75:    0.224s     143/s    -27.2%  p95:    0.227s     141/s    -28.9%
   8 GPUs:  p75:    0.350s      91/s    -95.4%  p95:    0.361s      88/s   -100.4%
  16 GPUs:  p75:    0.427s      74/s   -137.2%  p95:    0.441s      72/s   -134.2%

Benchmark: resnext50_32x4d with batch size 32

                  sec/iter    ex/sec      diff        sec/iter    ex/sec      diff
   1 GPUs:  p75:    0.158s     202/s     -7.3%  p95:    0.159s     201/s     -7.5%
   2 GPUs:  p75:    0.169s     189/s    -10.5%  p95:    0.170s     188/s    -11.0%
   4 GPUs:  p75:    0.176s     181/s    -14.8%  p95:    0.179s     178/s    -16.4%
   8 GPUs:  p75:    0.224s     142/s    -43.1%  p95:    0.232s     137/s    -47.8%
  16 GPUs:  p75:    0.268s     119/s    -70.2%  p95:    0.281s     113/s    -77.9%

Benchmark: resnext101_32x8d with batch size 32

                  sec/iter    ex/sec      diff        sec/iter    ex/sec      diff
   1 GPUs:  p75:    0.445s      71/s     -4.6%  p95:    0.446s      71/s     -4.7%
   2 GPUs:  p75:    0.485s      65/s    -10.7%  p95:    0.494s      64/s    -12.5%
   4 GPUs:  p75:    0.501s      63/s    -14.2%  p95:    0.505s      63/s    -14.8%
   8 GPUs:  p75:    0.734s      43/s    -64.0%  p95:    0.745s      42/s    -66.1%
  16 GPUs:  p75:    0.868s      36/s    -92.4%  p95:    0.877s      36/s    -94.2%
```

pietern committed:
Update on "Add distributed data parallel benchmark tool"


The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

Differential Revision: [D18781560](https://our.internmc.facebook.com/intern/diff/D18781560)

[ghstack-poisoned]

pietern committed:
Update on "Add distributed data parallel benchmark tool"


The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

Differential Revision: [D18781560](https://our.internmc.facebook.com/intern/diff/D18781560)

[ghstack-poisoned]

pietern commented:
@zhaojuanmao @pritamdamania87 Thanks for the reviews. This is ready for another pass.

zhaojuanmao reviewed (approved):
Thanks for adding this! It is really super useful!

pietern commented:
@pytorchbot rebase this please

pietern committed:
Rebase on "Add distributed data parallel benchmark tool"


The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

Differential Revision: [D18781560](https://our.internmc.facebook.com/intern/diff/D18781560)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 94eac9e5:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=94eac9e540a44c10d0468fe8208bbad71838aa90).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4358798) pytorch_linux_xenial_cuda10_1_cudnn7_py3_NO_AVX2_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92009399) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92009399)) :snowflake:
<details>
<summary>
<code>Jan 28 15:32:46 RuntimeError: quantization scale should be > 0</code>
</summary>

```
Jan 28 15:32:46   File "test_quantization.py", line 730, in test_quantized_rnn 
Jan 28 15:32:46     y, (h, c) = cell_dq(x, (h, c)) 
Jan 28 15:32:46   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 540, in __call__ 
Jan 28 15:32:46     result = self.forward(*input, **kwargs) 
Jan 28 15:32:46   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 373, in forward 
Jan 28 15:32:46     return self.forward_tensor(input, hx) 
Jan 28 15:32:46   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 334, in forward_tensor 
Jan 28 15:32:46     input, hx, batch_sizes, max_batch_size, sorted_indices) 
Jan 28 15:32:46   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 315, in forward_impl 
Jan 28 15:32:46     self.batch_first, dtype=self.dtype, use_dynamic=True) 
Jan 28 15:32:46 RuntimeError: quantization scale should be > 0 
Jan 28 15:32:46  
Jan 28 15:32:46 ---------------------------------------------------------------------- 
Jan 28 15:32:46 Ran 36 tests in 56.185s 
Jan 28 15:32:46  
Jan 28 15:32:46 FAILED (errors=1, skipped=1) 
Jan 28 15:32:46  
Jan 28 15:32:46 Generating XML reports... 
Jan 28 15:32:46 Traceback (most recent call last): 
Jan 28 15:32:46   File "test/run_test.py", line 456, in <module> 
Jan 28 15:32:46     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

pietern committed:
Fix lint on "Add distributed data parallel benchmark tool"


The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

Differential Revision: [D18781560](https://our.internmc.facebook.com/intern/diff/D18781560)

[ghstack-poisoned]

mrshenli commented:
@pytorchbot retest this please

mrshenli commented:
closing this PR as the same code is landed by #35198

jose-solorzano commented:
I guess I'm not understanding these results. Do they show it takes longer per iteration with more GPUs?
